In [4]:
import pandas as pd
import json

In [5]:
# Ajustar la opción para mostrar más caracteres en cada columna
pd.set_option('display.max_colwidth', None)

In [6]:
# Abrir el archivo JSON y leer línea por línea
with open(r'C:\\Users\\Pablo\\OneDrive\\Henry\\ProyectoFinal\\Google Maps\\metadata-sitios\\1.json') as file:
    data = [json.loads(line) for line in file]

# Convertir la lista de diccionarios en un DataFrame
df1 = pd.DataFrame(data)

In [ ]:
# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\2.json'
# Leer el JSON en un DataFrame
df2 = pd.read_json(ruta_json, lines=True)

# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\3.json'
# Leer el JSON en un DataFrame
df3 = pd.read_json(ruta_json, lines=True)

# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\4.json'
# Leer el JSON en un DataFrame
df4 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\5.json'
# Leer el JSON en un DataFrame
df5 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\6.json'
# Leer el JSON en un DataFrame
df6 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\7.json'
# Leer el JSON en un DataFrame
df7 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\8.json'
# Leer el JSON en un DataFrame
df8 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\9.json'
# Leer el JSON en un DataFrame
df9 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\10.json'
# Leer el JSON en un DataFrame
df10 = pd.read_json(ruta_json, lines=True)

In [7]:
def extractCity(df):
    """
    Esta función toma un DataFrame que contiene una columna llamada 'address', 
    la cual tiene direcciones formateadas con comas. La función extrae el valor
    que está entre la segunda y tercera coma (que representa la ciudad) y lo 
    almacena en una nueva columna llamada 'city'.

    Parámetros:
    df (DataFrame): El DataFrame que contiene la columna 'address'.

    Retorna:
    DataFrame: El DataFrame original con una nueva columna 'city' que contiene 
               el valor entre la segunda y tercera coma de 'address'.
    """
    
    # Dividir la columna 'address' en partes usando la coma como delimitador
    # Luego, extraer el segundo elemento (índice 2), que está entre la segunda y tercera coma

    df['city'] = df['address'].str.split(',').str[2].str.strip()
    
    return df

In [34]:
def nulosYVacios(df):
    """
    Este procedimiento recibe un DataFrame y muestra la cantidad de valores nulos, vacíos y la cantidad de 
    valores duplicados en todas las columnas excepto en las 6, 10, 11 y 13 del DataFrame. También muestra 
    el mínimo, máximo y media para columnas de tipo float e integer.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    """
    print("Informe de Nulos, Vacíos y Duplicados:")
    print("-" * 30)

    # Total de registros en el DataFrame
    total_registros = len(df)

    # Imprimir total de registros
    print(f"Total de registros: {total_registros}")
    print("-" * 30)

    # Iterar sobre todas las columnas para mostrar nulos y vacíos
    for columna in df.columns:
        # Contar valores nulos
        nulos = df[columna].isnull().sum()

        # Contar valores vacíos (considerando que solo se cuentan si son cadenas)
        vacios = (df[columna] == '').sum() if df[columna].dtype == 'object' else 0

        # Calcular porcentajes
        porcentaje_nulos = (nulos / total_registros) * 100 if total_registros > 0 else 0
        porcentaje_vacios = (vacios / total_registros) * 100 if total_registros > 0 else 0

        # Imprimir resultados
        print(f"Columna: {columna}")
        print(f"  Valores Nulos: {nulos}")
        print(f"  Valores Vacíos: {vacios}")
        print(f"  Porcentaje Nulos: {porcentaje_nulos:.2f}%")
        print(f"  Porcentaje Vacíos: {porcentaje_vacios:.2f}%")

        # Contar y mostrar la cantidad de valores duplicados solo si no es una de las columnas con datos anidados
        if columna not in [df.columns[6], df.columns[10], df.columns[11], df.columns[13]]:
            # Filtrar los valores no nulos antes de contar duplicados
            cantidad_duplicados = df[columna].dropna().duplicated(keep=False).sum()
            print(f"  Cantidad de valores duplicados (solo no nulos): {cantidad_duplicados}")

        # Mostrar mínimo, máximo y media si la columna es de tipo float o int
        if df[columna].dtype in ['float64', 'int64']:
            minimo = df[columna].min()
            maximo = df[columna].max()
            media = df[columna].mean()
            print(f"  Mínimo: {minimo}")
            print(f"  Máximo: {maximo}")
            print(f"  Media: {media:.2f}")

        print("-" * 30)


Análisis preliminar de calidad de datos

In [ ]:
# Obtener la ciudad donde se encuentra el sitio con reseña
extractCity(df1)
extractCity(df2)
extractCity(df3)
extractCity(df4)
extractCity(df5)
extractCity(df6)
extractCity(df7)
extractCity(df8)
extractCity(df9)
extractCity(df10)

In [9]:
def rankingFrecuenciaCiudad(df, x=10):
    """
    Este procedimiento recibe un DataFrame y un número entero x.
    Muestra en pantalla el ranking de los primeros x elementos con mayor frecuencia en la columna 'city'.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    x (int): El número de elementos a mostrar en el ranking.
    """
    print(f"Ranking de los primeros {x} elementos más frecuentes en la columna 'city':")
    print("-" * 60)

    # Verifica si la columna 'city' existe en el DataFrame
    if 'city' in df.columns:
        # Obtener la frecuencia de cada valor en la columna 'city'
        frecuencias = df['city'].value_counts()

        # Obtener los primeros x elementos con mayor frecuencia
        ranking = frecuencias.head(x)

        # Mostrar el ranking
        for index, value in ranking.items():
            print(f"{index}: {value}")
    else:
        print("La columna 'city' no existe en el DataFrame.")

    print("-" * 60)

In [51]:
def mostrarCantidadFilasDuplicadas(df):
    """
    Filtra las filas duplicadas en un DataFrame según columnas específicas,
    y muestra la cantidad de filas duplicadas en la salida de pantalla.
  
    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    """
    # Filtrar duplicados considerando las columnas específicas
    filas_duplicadas = df[df.duplicated(subset=[df.columns[i] for i in [0, 1, 2, 4, 5]], keep=False)]
    
    # Contar la cantidad de filas duplicadas
    cantidad_duplicados = len(filas_duplicadas)
    
    # Mostrar la cantidad de filas duplicadas
    print(f"Cantidad de filas duplicadas: {cantidad_duplicados}")

Frecuencia de recomendaciones por ciudad segun google Maps

In [12]:
# Obtener la frecuencia de cada valor en la columna 'city'
rankingFrecuenciaCiudad(df1)

Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 2599
New York: 1903
Chicago: 1799
Los Angeles: 1663
Brooklyn: 1550
San Antonio: 1200
Dallas: 1154
Miami: 1139
Las Vegas: 1089
Philadelphia: 1044
------------------------------------------------------------


In [ ]:
rankingFrecuenciaCiudad(df2)

In [ ]:
rankingFrecuenciaCiudad(df3)

In [33]:
rankingFrecuenciaCiudad(df4)
rankingFrecuenciaCiudad(df5)
rankingFrecuenciaCiudad(df6)
rankingFrecuenciaCiudad(df7)
rankingFrecuenciaCiudad(df8)
rankingFrecuenciaCiudad(df9)
rankingFrecuenciaCiudad(df10)

Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 2511
New York: 1855
Los Angeles: 1713
Chicago: 1679
Brooklyn: 1596
Miami: 1337
San Antonio: 1249
Las Vegas: 1229
Philadelphia: 1204
Dallas: 1143
------------------------------------------------------------
Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 2504
New York: 1825
Chicago: 1763
Los Angeles: 1625
Brooklyn: 1610
San Antonio: 1299
Dallas: 1185
Miami: 1148
Phoenix: 1071
Philadelphia: 1051
------------------------------------------------------------
Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 2598
Chicago: 1812
New York: 1810
Brooklyn: 1636
Los Angeles: 1401
San Antonio: 1366
Dallas: 1288
Miami: 1232
Las Vegas: 1172
Phoenix: 1119
------------------

In [35]:
nulosYVacios(df1)

Informe de Nulos, Vacíos y Duplicados:
------------------------------
Total de registros: 275001
------------------------------
Columna: name
  Valores Nulos: 7
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 84218
------------------------------
Columna: address
  Valores Nulos: 10062
  Valores Vacíos: 0
  Porcentaje Nulos: 3.66%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 51265
------------------------------
Columna: gmap_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 53146
------------------------------
Columna: description
  Valores Nulos: 261846
  Valores Vacíos: 0
  Porcentaje Nulos: 95.22%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 10123
------------------------------
Columna: latitude
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 

In [38]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 275001 entries, 0 to 275000
Data columns (total 16 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   name              274994 non-null  object 
 1   address           264939 non-null  object 
 2   gmap_id           275001 non-null  object 
 3   description       13155 non-null   object 
 4   latitude          275001 non-null  float64
 5   longitude         275001 non-null  float64
 6   category          272740 non-null  object 
 7   avg_rating        275001 non-null  float64
 8   num_of_reviews    275001 non-null  int64  
 9   price             13450 non-null   object 
 10  hours             192448 non-null  object 
 11  MISC              194972 non-null  object 
 12  state             195523 non-null  object 
 13  relative_results  238771 non-null  object 
 14  url               275001 non-null  object 
 15  city              261516 non-null  object 
dtypes: float64(3), int64

Puede ser que se encuentren valores repetidos de lugar en el mapa, porque quizas es un edificio, o antes estuvo otro, por lo tanto habra que ver si es que tienen el mismo nombre. coincidencia entre esas coolumnas.
 0   name              274994 non-null  object 
 1   address           264939 non-null  object 
 2   gmap_id           275001 non-null  object 
 4   latitude          275001 non-null  float64
 5   longitude         275001 non-null  float64
 14  url               275001 non-null  object 

In [52]:
# Filas duplicas considerando las columnas 0 a 5
mostrarCantidadFilasDuplicadas(df1)

Cantidad de filas duplicadas: 53146


In [135]:
# Cargar el archivo de business
dfYeplp = pd.read_pickle(r'C:\Users\Pablo\OneDrive\Henry\Proyecto Final\Yelp\business.pkl')

In [152]:
dfYeplp.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150346 entries, 0 to 150345
Data columns (total 28 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   business_id   150346 non-null  object
 1   name          150346 non-null  object
 2   address       150346 non-null  object
 3   city          150346 non-null  object
 4   state         150343 non-null  object
 5   postal_code   150346 non-null  object
 6   latitude      150346 non-null  object
 7   longitude     150346 non-null  object
 8   stars         150346 non-null  object
 9   review_count  150346 non-null  object
 10  is_open       150346 non-null  object
 11  attributes    136602 non-null  object
 12  categories    150243 non-null  object
 13  hours         127123 non-null  object
 14  business_id   5 non-null       object
 15  name          5 non-null       object
 16  address       5 non-null       object
 17  city          5 non-null       object
 18  state         5 non-null     

In [ ]:
# Filtrar las filas donde gmap_id está duplicado
filas_duplicadas = df1[df1['gmap_id'].duplicated(keep=False)]


In [ ]:
# Ordenar el DataFrame por la columna gmap_id
df_ordenado = filas_duplicadas.sort_values(by='gmap_id')


df_ordenado


In [27]:
df1.iloc[:,[0,1,2,3,14]]

name  \
0                       Porter Pharmacy   
1                          City Textile   
2                          San Soo Dang   
3                          Nova Fabrics   
4                      Nobel Textile Co   
...                                 ...   
274996                     Maven Realty   
274997               Stone Castle Fence   
274998  Health Insurance Solutions Team   
274999          Bernard Building Center   
275000    HMR Motor Works and Auto Care   

                                                                       address  \
0                          Porter Pharmacy, 129 N Second St, Cochran, GA 31014   
1                        City Textile, 3001 E Pico Blvd, Los Angeles, CA 90023   
2                       San Soo Dang, 761 S Vermont Ave, Los Angeles, CA 90005   
3                          Nova Fabrics, 2200 E 11th St, Los Angeles, CA 90021   
4                        Nobel Textile Co, 719 E 9th St, Los Angeles, CA 90021   
...                                                                        ...   
274996      Maven Realty, 2180 McMillan Ave #71672, North Charleston, SC 29405   
274997     Stone Castle Fence, 124 Cypress Gardens Rd, Moncks Corner, SC 29461   
274998  Health Insurance Solutions Team, 6650 Rivers Ave, Charleston, SC 29406   
274999            Bernard Building Center, 27300 M-32, M-32, Hillman, MI 49746   
275000                            HMR Motor Works and Auto Care, Mio, MI 48647   

                                      gmap_id description  \
0       0x88f16e41928ff687:0x883dad4fd048e8f8        None   
1       0x80c2c98c0e3c16fd:0x29ec8a728764fdf9        None   
2       0x80c2c778e3b73d33:0xbdc58662a4a97d49        None   
3        0x80c2c89923b27a41:0x32041559418d447        None   
4       0x80c2c632f933b073:0xc31785961fe826a6        None   
...                                       ...         ...   
274996  0x88fe64e7daceaaab:0x101c046dcadcc9c3        None   
274997  0x88fe59f4602c95fd:0x57e2b1904d9e6949        None   
274998  0x88fe71f9617c15e1:0x89518c1d251706bf        None   
274999  0x4d34c7517d6e6e29:0x7fe553ee2f090cf0        None   
275000  0x88208f4360c053ab:0xe3d6883f19601a01        None   

                                                                                                                     url  
0       https://www.google.com/maps/place//data=!4m2!3m1!1s0x88f16e41928ff687:0x883dad4fd048e8f8?authuser=-1&hl=en&gl=us  
1       https://www.google.com/maps/place//data=!4m2!3m1!1s0x80c2c98c0e3c16fd:0x29ec8a728764fdf9?authuser=-1&hl=en&gl=us  
2       https://www.google.com/maps/place//data=!4m2!3m1!1s0x80c2c778e3b73d33:0xbdc58662a4a97d49?authuser=-1&hl=en&gl=us  
3        https://www.google.com/maps/place//data=!4m2!3m1!1s0x80c2c89923b27a41:0x32041559418d447?authuser=-1&hl=en&gl=us  
4       https://www.google.com/maps/place//data=!4m2!3m1!1s0x80c2c632f933b073:0xc31785961fe826a6?authuser=-1&hl=en&gl=us  
...                                                                                                                  ...  
274996  https://www.google.com/maps/place//data=!4m2!3m1!1s0x88fe64e7daceaaab:0x101c046dcadcc9c3?authuser=-1&hl=en&gl=us  
274997  https://www.google.com/maps/place//data=!4m2!3m1!1s0x88fe59f4602c95fd:0x57e2b1904d9e6949?authuser=-1&hl=en&gl=us  
274998  https://www.google.com/maps/place//data=!4m2!3m1!1s0x88fe71f9617c15e1:0x89518c1d251706bf?authuser=-1&hl=en&gl=us  
274999  https://www.google.com/maps/place//data=!4m2!3m1!1s0x4d34c7517d6e6e29:0x7fe553ee2f090cf0?authuser=-1&hl=en&gl=us  
275000  https://www.google.com/maps/place//data=!4m2!3m1!1s0x88208f4360c053ab:0xe3d6883f19601a01?authuser=-1&hl=en&gl=us  

[275001 rows x 5 columns]

In [153]:
# Crear un nuevo DataFrame usando loc
dfYeplp_copia = dfYeplp.copy()

Eliminar columnas duplicadas, desde la 14 a 27

In [154]:
# Crear copia donde solo se excluyen las duplicadas.
dfYeplp_copia = dfYeplp_copia.iloc[:,0:14]

In [155]:
dfYeplp_copia.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150346 entries, 0 to 150345
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   business_id   150346 non-null  object
 1   name          150346 non-null  object
 2   address       150346 non-null  object
 3   city          150346 non-null  object
 4   state         150343 non-null  object
 5   postal_code   150346 non-null  object
 6   latitude      150346 non-null  object
 7   longitude     150346 non-null  object
 8   stars         150346 non-null  object
 9   review_count  150346 non-null  object
 10  is_open       150346 non-null  object
 11  attributes    136602 non-null  object
 12  categories    150243 non-null  object
 13  hours         127123 non-null  object
dtypes: object(14)
memory usage: 17.2+ MB


Ranking de ciudades con mas comentarios segun Yelp

In [1]:
rankingFrecuenciaCiudad(dfYeplp_copia, 5)

NameError: name 'rankingFrecuenciaCiudad' is not defined

In [8]:
# Eliminar las columnas 'url', 'relative_results', 'state' y 'price'
df_cleaned = df1.drop(columns=['url', 'relative_results', 'state', 'price'])

# Ver las primeras filas del DataFrame limpio
print(df_cleaned.head())

               name                                            address  \
0   Porter Pharmacy  Porter Pharmacy, 129 N Second St, Cochran, GA ...   
1      City Textile  City Textile, 3001 E Pico Blvd, Los Angeles, C...   
2      San Soo Dang  San Soo Dang, 761 S Vermont Ave, Los Angeles, ...   
3      Nova Fabrics  Nova Fabrics, 2200 E 11th St, Los Angeles, CA ...   
4  Nobel Textile Co  Nobel Textile Co, 719 E 9th St, Los Angeles, C...   

                                 gmap_id description   latitude   longitude  \
0  0x88f16e41928ff687:0x883dad4fd048e8f8        None  32.388300  -83.357100   
1  0x80c2c98c0e3c16fd:0x29ec8a728764fdf9        None  34.018891 -118.215290   
2  0x80c2c778e3b73d33:0xbdc58662a4a97d49        None  34.058092 -118.292130   
3   0x80c2c89923b27a41:0x32041559418d447        None  34.023669 -118.232930   
4  0x80c2c632f933b073:0xc31785961fe826a6        None  34.036694 -118.249421   

              category  avg_rating  num_of_reviews  \
0           [Pharmacy]    

In [13]:
print(df_cleaned['MISC'][0])


{'Service options': ['In-store shopping', 'Same-day delivery'], 'Health & safety': ['Mask required', 'Staff required to disinfect surfaces between visits'], 'Accessibility': ['Wheelchair accessible entrance'], 'Planning': ['Quick visit']}


In [19]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 275001 entries, 0 to 275000
Data columns (total 11 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   name            274994 non-null  object 
 1   address         264939 non-null  object 
 2   gmap_id         275001 non-null  object 
 3   description     13155 non-null   object 
 4   latitude        275001 non-null  float64
 5   longitude       275001 non-null  float64
 6   category        272740 non-null  object 
 7   avg_rating      275001 non-null  float64
 8   num_of_reviews  275001 non-null  int64  
 9   hours           192448 non-null  object 
 10  MISC            194972 non-null  object 
dtypes: float64(3), int64(1), object(7)
memory usage: 23.1+ MB
